File: lib/md2pdf/title_page.lua
-- Wrap the H1 + the leading lead blocks in
-- <div class="title-page"> so CSS can vertically center
-- them on the first page. The wrapping kicks in only
-- when the document gets a TOC and has no explicit
-- ::: intro ::: block. In every other case the
-- title and lead content flow normally from the top of
-- page 1, exactly as plain pandoc renders them.
--
-- Rule:
--   wrap when: TOC enabled AND no Div { class = "intro" }
--   else:      no wrap

local function is_intro(block)
  if block.t ~= 'Div' then return false end
  for _, c in ipairs(block.classes or {}) do
    if c == 'intro' then return true end
  end
  return false
end

function Pandoc(doc)
  local toc =
    pandoc.utils.stringify(doc.meta['md2pdf-toc'] or '') == 'true'
  if not toc then return doc end

  for _, b in ipairs(doc.blocks) do
    if is_intro(b) then return doc end
  end

  local h1_idx
  for i, b in ipairs(doc.blocks) do
    if b.t == 'Header' and b.level == 1 then
      h1_idx = i
      break
    end
  end
  if not h1_idx then return doc end

  local end_idx = #doc.blocks + 1
  for i = h1_idx + 1, #doc.blocks do
    local b = doc.blocks[i]
    if b.t == 'Header' and b.level == 2 then
      end_idx = i
      break
    end
  end

  local wrapped = {}
  for i = h1_idx, end_idx - 1 do
    table.insert(wrapped, doc.blocks[i])
  end
  local div = pandoc.Div(wrapped, pandoc.Attr('', { 'title-page' }, {}))

  local new_blocks = {}
  for i = 1, h1_idx - 1 do
    table.insert(new_blocks, doc.blocks[i])
  end
  table.insert(new_blocks, div)
  for i = end_idx, #doc.blocks do
    table.insert(new_blocks, doc.blocks[i])
  end
  doc.blocks = new_blocks
  return doc
end

File: lib/transpareo/md2pdf/filters/title_page.rb
# frozen_string_literal: true

module Transpareo
  module Md2pdf
    module Filters
      # Wraps the H1 and the lead blocks that follow it in
      # <div class="title-page"> so CSS can centre them vertically
      # on page 1.
      #
      # This only applies when the document gets a TOC and has no
      # explicit `::: intro :::` block. In every other case the
      # title flows from the top of page 1 instead.
      module TitlePage
        module_function

        def call(doc)
          return unless doc[:toc]
          return if doc.fragment.at_css('div.intro')

          h1 = doc.fragment.at_css('h1')
          return unless h1

          lead = collect_lead(h1)
          wrapper = doc.build('<div class="title-page"></div>')
            .at_css('div')

          h1.add_previous_sibling(wrapper)
          lead.each { |node| wrapper.add_child(node) }
        end

        # The H1 plus every following sibling up to the first H2.
        def collect_lead(title)
          nodes = [title]
          node = title.next_sibling
          while node
            break if node.element? && node.name == 'h2'

            nodes << node
            node = node.next_sibling
          end
          nodes
        end
      end
    end
  end
end
